### ISSUE TITLE
iOS keyboard covers inputs when using useDialog/useModalOverlay

### ISSUE THREAD

**Issue by JRS-Developer** [adobe/react-spectrum#5926]
Additional information: https://github.com/adobe/react-spectrum/issues/6405 shows that this is not only limited to full screen dialogs, but modal experiences in general.

### Provide a general summary of the issue here

we noticed that using the example from https://react-spectrum.adobe.com/react-aria/useDialog.html causes issues when creating a full screen dialog where an input is at the bottom of the dialog, it's impossible to see the input, the keyboard covers it, it only happens on iOS, old and new safari versions. we tried using radix dialog and it works properly so it seems some internal logic of react-aria.

https://github.com/adobe/react-spectrum/assets/77207702/9e493c13-bc9d-499a-af16-5a5817f3f71c

### 🤔 Expected Behavior?

the dialog should leave the browser to focus the input and "fix" by itself the keyboard covering.

### 😯 Current Behavior

the keyboard covers the input and it's impossible to scroll down, the dialog scrolls up when trying it. 

### 💁 Possible Solution

no sure

### 🖥️ Steps to Reproduce

1. Open this [codesandbox](https://codesandbox.io/p/github/JRS-Developer/testing-dialogs/main?layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clsxs3ow500073d5tx2jr1ii8%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clsxs3ow500023d5tvzqvngt1%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clsxs3ow500043d5tos3c04kt%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clsxs3ow500063d5tsdxj7c0g%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clsxs3ow500023d5tvzqvngt1%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsxs3ow400013d5tag68uvix%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%257D%255D%252C%2522id%2522%253A%2522clsxs3ow500023d5tvzqvngt1%2522%252C%2522activeTabId%2522%253A%2522clsxs3ow400013d5tag68uvix%2522%257D%252C%2522clsxs3ow500063d5tsdxj7c0g%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsxs3ow500053d5t4sjwoe1w%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clsxs3ow500063d5tsdxj7c0g%2522%252C%2522activeTabId%2522%253A%2522clsxs3ow500053d5t4sjwoe1w%2522%257D%252C%2522clsxs3ow500043d5tos3c04kt%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clsxs3ow500033d5ts6uhp2bi%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522id%2522%253A%2522clsxs3ow500043d5tos3c04kt%2522%252C%2522activeTabId%2522%253A%2522clsxs3ow500033d5ts6uhp2bi%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D), it uses react aria and radix, same styling.
2. Click on Open React aria dialog and select input at the bottom of the dialog and the keyboard will cover the input and it won't zoom correctly.
3. Try the same with Radix and it won't happen

### Version

3.32.1

### What browsers are you seeing the problem on?

Safari

### What operating system are you using?

iOS

**Comment by LFDanLu** [comment]
Since you are using a hook implementation, you'll probably want to use [useViewportSize](https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useViewportSize.ts) to determine the height of your modal for cases like these. Alternatively, you could use React Aria Components for your Modal which will provide the viewport's size, see https://react-spectrum.adobe.com/react-aria/Modal.html#modaloverlay

**Comment by JRS-Developer** [comment]
Use the height of the modal for? Could you explain more? @LFDanLu

**Comment by LFDanLu** [comment]
@JRS-Developer since the virtual keyboard takes up space in the viewport, you can set the height of your fullscreen modal to match the height of the the visual viewport and thus have it adapt to when the virtual keyboard is open. You can see this in action with the edit modal in the first Table example on the [React Aria home page](https://react-spectrum.adobe.com/react-aria/index.html). The style in question: https://github.com/adobe/react-spectrum/blob/fa3862eab668b43459dd99f7fd4bf74e6eda492b/packages/dev/docs/pages/react-aria/home/ExampleApp.tsx#L488

**Comment by JRS-Developer** [comment]
Hello, will this work in cases where the modal is bigger that the mobile screen? Like an modal that overflows? Like the scroll=body of Mui? https://mui.com/material-ui/react-dialog/#scrolling-long-content

**Comment by LFDanLu** [comment]
The `scroll=body` case shown for material-ui probably wouldn't work since the strategy I suggested above is to make sure that the modal's height is equal to the visual viewport's height and to make the body of the modal scroll instead.

**Comment by JRS-Developer** [comment]
@LFDanLu Yeah, we use a lot dialogs that work like the `scroll=body` and it's pretty anoying for IOS users

**Comment by LFDanLu** [comment]
I would need to dig further into what radix is doing vs our implementation, but it honestly feels to me to be a pretty odd experience in that case too. From your video, the virtual keyboard still covers the textfield when opened until after some delay (not sure what caused it to scroll into view). If an additional interaction was required to make the input scroll into view after the virtual keyboard opened then that feels less than ideal. With the example in the React Aria home page, the modal shrinks and the input remains in view:

https://github.com/adobe/react-spectrum/assets/9661127/561b8b6c-f45f-4d19-ac02-75830fbd8924

**Comment by JRS-Developer** [comment]
Hello, I updated the example with a really simple dialog without any library and it works as radix, it's really an issue with react-aria, here is the video using all dialogs including the simple one

https://github.com/adobe/react-spectrum/assets/77207702/6e951feb-f395-451a-aea6-61fa66a05fa9

**Comment by JRS-Developer** [comment]
Maybe some logic inside Overlay or useModalOverlay, no really sure

**Comment by LFDanLu** [comment]
For whoever picks up this bug in the future, note that the same issue can exhibit itself for our fullscreen modal dialog in RSP as well, looks like either an issue with the scrollIntoView code calculation or a race condition between when the scrollIntoView code calls itself and when the modal adjusts its own height to match the new visual viewport height

**Comment by IonelLupu** [comment]
I also tested in a simulator and even if it's not similar to a real device, there is some similar behaviour between the two. Focusing on date elements in the date selector makes the UI jump up and down as seen in the last part of the video: 
https://github.com/adobe/react-spectrum/assets/4083652/4b38d902-30ed-49da-9ff1-8da0c5abb6a3

**Comment by brianudensi** [comment]
Similar issue is also observed when using the `useModalOverlay` hook to create a modal. As seen [here](https://github.com/adobe/react-spectrum/issues/6405)

**Comment by brianudensi** [comment]
@LFDanLu thanks for the switch response, just curious on when we should be looking out for an update regarding this issue?

**Comment by LFDanLu** [comment]
Hard for me to say, the team is pretty booked for the next couple of sprints, but I can see about bring this up in our next planning session and see if we can get it slotted in. Any help investigating solutions is welcome!

**Comment by thebuilder** [comment]
Running into the same issue - Have implemented https://react-spectrum.adobe.com/react-aria/examples/framer-modal-sheet.html and was afraid it was related to `transform` from Framer Motion. But after removing all styling on the modal/dialog it still happens. It's definitely related to the scroll prevention. You can see it tries to scroll the screen for a split second, before snapping back.

![RPReplay_Final1720000461](https://github.com/adobe/react-spectrum/assets/3764345/69c313ae-c788-4fea-b381-6ea23270af8d)

**Comment by subvertallchris** [comment]
So very sorry to be that guy but any chance this made it into the queue @LFDanLu ?

**Comment by LFDanLu** [comment]
It is still in our backlog unfortunately.

**Comment by Emiliano-Bucci** [comment]
Great but this is a blocking issue, i mean..

**Comment by LFDanLu** [comment]
I can see about bringing it up again for our upcoming sprint planning. If you could thumbs up the issue, that would be helpful for us to gauge priority.

**Comment by devongovett** [comment]
This is a free and open source library, provided [as is](https://github.com/adobe/react-spectrum/blob/10a469d547bb8273f78aa220682c7022796a6268/LICENSE#L143-L151). We aren't paid to work on your personal priorities. You are welcome to fork the repo, use [patch-package](https://www.npmjs.com/package/patch-package), or contribute a fix if it is blocking you, otherwise, we will try to incorporate it into our upcoming work. Thanks for your patience.

**Comment by devongovett** [comment]
Here's a fork of the original sandbox that works for me: https://codesandbox.io/p/github/JRS-Developer/testing-dialogs/csb-jh5gwz/draft/nameless-sea See demo here on iOS: https://jh5gwz-3000.csb.app/.

The only change I made was to set a height on the `position: fixed` element that uses the visual viewport height (the space above the keyboard) instead of setting `bottom: 0`. If you're using hooks, you could do it like this:

```js
height: useViewportSize().height + "px",
```

If you're using React Aria Components, this is available as the `--visual-viewport-height` custom property. The [example in our docs](https://react-spectrum.adobe.com/react-aria/Modal.html) also uses this in its styles, and appears to work correctly on iOS.

We expect that the dialog will be positioned above the keyboard, and scroll within that area, rather than extending behind the keyboard. This needs to be handled in your styles.

I noticed that the other examples in the sandbox (radix and simple dialog) scroll the input into view initially, but you can't actually scroll up to the top of the dialog anymore once the keyboard is visible. That's because the dialog gets pushed up, and it is too tall so extends outside the viewport. I think it's a better experience to shrink the height of the dialog when the keyboard is open, so it is scrollable above the keyboard and all content is reachable.

Given this is working with the above tweak, it seems like this isn't so much a bug as something we could improve in the docs. Am I missing something else that's broken?

**Comment by Emiliano-Bucci** [comment]
@devongovett The fact that you work free on this doesn't give you the right to do shitty stuff for years and then justify like: "hey, no one is paying me to do that, not my fault". Well if no one is paying you to do it, or don't do it or if you want to do it then do it right, and by do it right i mean that if a FUCKING CRITICAL issue raise, you must fix it. This stupid open source philosofy that "since no one is paying me I have the right to destroy the entire internet" is cancer.

**Comment by Emiliano-Bucci** [comment]
@snowystinger I see that you've deleted my comment. I want you to understand that I'm not frustrated at all, because I just simply resolved my problem by using another library that work as expected without spending time trying to make things work because people is not able to do it's job right. I was just saying that the fact that you work free doesn't give you the **right** to break stuff in the internet, that's not a good philosofy in life :)

**Comment by thebuilder** [comment]
Thanks for the solution @devongovett - Much appreciated. It also seemed like a weird oversight given all the work you have put into making everything in this library work in tandem.

It would be good to clarify the the importance of the visual viewport in docs/examples, since it's not super clear what it actually does.

The [Modal Sheet](https://react-spectrum.adobe.com/react-aria/examples/framer-modal-sheet.html) example could use a documentation tweak, so it also includes the visual viewportsize. It's a bit more tricky here since it adds Framer Motion, that relies on the calculated height. Right now it just reads `window.innerHeight`.
